woocommerce/woocommerce-gutenberg-products-block
# Generate block pages for tests.

**TIL:** This PR changes the way we insert and create blocks in our tests, instead of inserting the block each time and comparing it with a snapshot, a very strict process, overlay strict, that causes issues, it will attempt to open a page with that block already there and confirm that it loads fine, so it delegates the validation to Gutenberg, the thing we can about in the first place.

### What is introduced:

- a new alias package `@woocommerce/blocks-test-utils` that will hold all of our test utils, a  growing collection (in other branches at least).
- an E2E utility function `visitBlockPage`, that takes a string, in the shape of `BLOCK_NAME Block`, it will attempt to:
    - Search for that page.
    - If it exists, it will visit it.
    - If it doesn’t exist, it will:
        - Create a new page.
        - Inserts that block.
        - Get the page content.
        - Generate a new `block-name.fixture.json` that contains the title and content.
        - Save it on `__fixtures__` folder on the same level as the test that caused it.
        - Block Fixtures are treated like snapshots and should committed.
- Expand our fixture loaders with two functions `createBlockPages` and `deleteBlockPages`.
    - `createBlockPages` will loop our `spec` folder (the place we hold our E2E tests in), searching for `*.fixture.json` files, and generating pages preloaded with that content.
    - `deleteBlockPages` will take an array of pages IDs and delete them (to keep tests clean).
- Refactor our existing E2E tests in some ways  (that I hope to abstract further):
    - All tests have a `block` object containing the block title, slug, and classname, those are helpful to copy paste code between tests.
    - All tests follow the pattern of:
        1. Calling  `visitBlockPage` beforeAll.
        2. Ensure the block renders fine (no Gutenberg crash, no ErrorBoundary crash, no validation crash, and the block actually rendering), I will later abstract those to a jest matcher, so basically `await expect( page ).toRenderBlock( block );`
        3. Ensure test can be inserted more than once or only once.
    - Entering the block page happens once per block, previously it was once per test, **this is important**, as your test will break if another test change something, take `product-search`, one test will hide the title, the next test will try to edit the title text but it will break because the title is hidden, **you should always remember to reset the changes your test does.**
I went with this because it drastically reduces our test duration, each test takes less than 500ms to be done, if we go all the process, it will take up to 5 seconds.

with `beforeEach`
```
  All Products Block
    ✓ can only be inserted once (4054ms)
    ✓ renders without crashing (2432ms)
```

with `beforeAll`
```
 PASS  tests/e2e-tests/specs/backend/all-products.test.js (7.433s)
  All Products Block
    ✓ can only be inserted once (374ms)
    ✓ renders without crashing (172ms)
```

Now it might not seem like much, we still have some manual product import, and few tests, but with two other branches coming, we will test 28 block, with 5 or 6 tests per block, spanning 4 WordPress configurations.

### How to test

- Tests should pass in travis.
- Locally you could try running `npm run test:e2e` after starting docker with `npm run test:e2e:up:build` give it some time to be fully up.
- You can try to delete the `__fixtures__` and running the tests to see if it was generated.
- You can see a test visually by running `npm run test:e2e-dev PATH_TO_TEST`.

### Follows ups:

- Create a jest matcher to abstract some tests. #2709
- Document the usage of those patterns, including how to write E2E tests and reasoning behind them.
- Write a P2 of how we write those tests.
- Document the Fixture pattern.
